File: Java/Ema/Src/main/java/com/thomsonreuters/ema/access/OmmQos.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2015. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;

/**
 * OmmQos represents Quality Of Service information in Omm.
 */
public interface OmmQos extends Data
{
	/**
	 * Rate represents Qos rate.
	 */
	public static class  Rate
	{
		/**
		 * Indicates tick by tick rate
		 */
		public final static int TICK_BY_TICK = 0;

		/**
		 * Indicates just in time conflated rate
		 */
		public final static int JUST_IN_TIME_CONFLATED = 0xFFFFFF00;
	}

	/**
	 * Timeliness represents Qos timeliness.
	*/
	public static class  Timeliness
	{
		/**
		 * Indicates real time timeliness
		 */
		public final static int REALTIME = 0;

		/**
		 * Indicates timeliness with an unknown delay value
		 */
		public final static int INEXACT_DELAYED = 0xFFFFFFFF;
	}

	/**
	 * Returns the QosRate value as a string format.
	 * 
	 * @return string representation of this object Rate
	*/
	public String rateAsString();
		
	/**
	 * Returns the QosTimeliness value as a string format.
	 * 
	 * @return string representation of this object timeliness
	*/
	public String timelinessAsString();

	/**
	 * Returns Timeliness.
	 * 
	 * @return value of OmmQos Timeliness
	 */
	public int timeliness();

	/** Returns Rate.
		@return value of OmmQos Rate
	*/
	public int rate();
}

File: Cpp-C/Ema/Src/Access/Impl/OmmQos.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright Thomson Reuters 2015. All rights reserved.            --
 *|-----------------------------------------------------------------------------
 */

#include "OmmQos.h"
#include "OmmQosDecoder.h"
#include "Utilities.h"
#include "ExceptionTranslator.h"
#include <new>

using namespace thomsonreuters::ema::access;

const EmaString TickByTickRateName( "TickByTick" );
const EmaString JustInTimeConflatedRateName( "JustInTimeConflated" );
EmaString UnknownQosRateName;

const EmaString RealTimeTimelinessName( "RealTime" );
const EmaString InexactDelayedTimelinessName( "InexactDelayed" );
EmaString UnknownQosTimelinessName;

OmmQos::OmmQos() :
 _pDecoder( new ( _space ) OmmQosDecoder() )
{
}

OmmQos::~OmmQos()
{
	_pDecoder->~OmmQosDecoder();
}

const EmaString& OmmQos::getRateAsString() const
{
	switch ( getRate() )
	{
	case TickByTickEnum :
		return TickByTickRateName;
	case JustInTimeConflatedEnum :
		return JustInTimeConflatedRateName;
	default :
		return UnknownQosRateName.set( "Rate: " ).append( (UInt64)getRate() );
	}
}

const EmaString& OmmQos::getTimelinessAsString() const
{
	switch ( getTimeliness() )
	{
	case RealTimeEnum :
		return RealTimeTimelinessName;
	case InexactDelayedEnum :
		return InexactDelayedTimelinessName;
	default :
		return UnknownQosTimelinessName.set( "Timeliness: " ).append( (UInt64)getTimeliness() );
	}
}

DataType::DataTypeEnum OmmQos::getDataType() const
{
	return DataType::QosEnum;
}

Data::DataCode OmmQos::getCode() const
{
	return _pDecoder->getCode();
}

UInt32 OmmQos::getTimeliness() const
{
	return _pDecoder->getTimeliness();
}

UInt32 OmmQos::getRate() const
{
	return _pDecoder->getRate();
}

const EmaString& OmmQos::toString() const
{
	return _pDecoder->toString();
}

const EmaString& OmmQos::toString( UInt64 ) const
{
	return _pDecoder->toString();
}

const EmaBuffer& OmmQos::getAsHex() const
{
	return _pDecoder->getHexBuffer();
}

Decoder& OmmQos::getDecoder()
{
	return *_pDecoder;
}

bool OmmQos::hasDecoder() const
{
	return true;
}

const Encoder& OmmQos::getEncoder() const
{
	return *static_cast<const Encoder*>( 0 );
}

bool OmmQos::hasEncoder() const
{
	return false;
}
